kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -303,7 +298,7 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
          *    It should be last so all callbacks observe consistent state
          *    of the job which doesn't depend on callback scheduling.
          */
-        onCompletionInternal(update, mode, suppressed)
+        onCompletionInternal(update, mode)
```

Comment:
It looks like this is the wrong place to handle an exception:
```
runBlocking {
  val job = Job()
  val ch = produce<Int>(job + Dispatchers.Default + CoroutineExceptionHandler {
         _, t -> Thread.sleep(100); println("Handled $t") }) {
     throw AssertionError()
  }
   ch.cancel()
   job.complete()
   job.join()
   println("Joined")
}
```

prints 
```
Joined
Handled java.lang.AssertionError
```

When completion handlers are invoked, it is already too late to handle exceptions

Reply:
Note that now we have unused `onCancellationInternal` that looks like the right place to handle exceptions in produce and actor

Reply:
...looking at `onCompletionInternal` that eventually leads to `onCancelled` or `onCompleted`

Both functions have a meaningful side-effect (future completion, exception handling, flux completion etc.).
It is inconsistent to have successful `join` while coroutine is still doing meaningful work that changes observing state.

Reply:
Answering in pieces. `onCancellation` is invoked too early, when the coroutine is just cancelled before we know its final exception. It is too early to **handle** exception there, because the exception is not final yet (coroutine might crash while handling it). The only legal use of `onCancellation` is to cancel whatever other work the coroutine might have been doing. Note, that `onCancellation` is always called _before_ cancelling all children. And it _is_ actually used in _actor_ builder to cancel the channel as soon as possible.

Reply:
Good catch with code example. I've further tweaked the order to resolve this problem. I think I made a break-through in understanding the problem: different kinds of coroutines need to plug into a different phases of "completion" processing:

* Various futures, produce, etc need to update their state even _before_ the state of the Job becomes final (to be race-free), so that a successful `join` always sees a "linked" object is a proper state. They all override `onCompletionInternal` (via `onCancelled`/`onCompleted`) to do so.
So far, I'm giving this guarantee only for failures, but we might consider extending this guarantee to successful completions, too (not sure we need to at the moment)

* Scoped coroutines need to resume the rest of the code _after_ everything is done. I've created a separate `afterCompletionInternal` for them (they also need to know `mode`, while `onCompletionInternal` does not need `mode`).